File: backend/pricing/schedule/schedule_curves.py
"""Curvas físico-financeiras mensais — espelho de frontend/lib/schedule-curves.ts."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pricing.models.budget_item import BudgetItem
    from pricing.schedule.schedule_models import ProjectSchedule, ScheduleTask


@dataclass
class MonthBucket:
    month_index: int
    label: str
    month_start_iso: str
    month_end_iso: str
    physical_monthly_pct: float
    physical_cumulative_pct: float
    financial_monthly: float
    financial_cumulative: float


def _parse_date(iso: str) -> date:
    y, m, d = iso[:10].split("-")
    return date(int(y), int(m), int(d))


def _days_between(start: str, end: str) -> int:
    a = _parse_date(start)
    b = _parse_date(end)
    return (b - a).days


def _add_days(iso: str, days: int) -> str:
    d = _parse_date(iso) + timedelta(days=days)
    return d.isoformat()[:10]


def _overlap_days(range_start: str, range_end: str, period_start: str, period_end: str) -> int:
    rs = _parse_date(range_start)
    re = _parse_date(range_end)
    ps = _parse_date(period_start)
    pe = _parse_date(period_end)
    start = max(rs, ps)
    end = min(re, pe)
    if end < start:
        return 0
    return (end - start).days + 1


def _month_start_date(project_start: str, month_index: int) -> date:
    s = _parse_date(project_start)
    month = s.month - 1 + month_index
    year = s.year + month // 12
    month = month % 12 + 1
    return date(year, month, 1)


def _month_end_date(d: date) -> date:
    if d.month == 12:
        return date(d.year + 1, 1, 1) - timedelta(days=1)
    return date(d.year, d.month + 1, 1) - timedelta(days=1)


def _format_iso(d: date) -> str:
    return d.isoformat()[:10]


def count_months(project_start: str, project_end: str) -> int:
    s = _parse_date(project_start)
    e = _parse_date(project_end)
    return max(1, (e.year - s.year) * 12 + (e.month - s.month) + 1)


def month_label(d: date) -> str:
    months = ("jan.", "fev.", "mar.", "abr.", "mai.", "jun.", "jul.", "ago.", "set.", "out.", "nov.", "dez.")
    return f"{months[d.month - 1]} de {d.year % 100:02d}"


def _clip_to_project(
    period_start: str,
    period_end: str,
    project_start: str,
    project_end: str,
) -> tuple[str, str] | None:
    end = project_end or project_start
    start = period_start if period_start >= project_start else project_start
    finish = period_end if period_end <= end else end
    if start > finish:
        return None
    return start, finish


def _row_for_task(rows: list[BudgetItem], task: ScheduleTask) -> BudgetItem | None:
    for row in rows:
        if row.row_id == task.budget_row_id:
            return row
    return None


def _leaf_task_for_service(schedule: ProjectSchedule, row_id: str) -> ScheduleTask | None:
    for task in schedule.tasks:
        if (
            task.budget_row_id == row_id
            and not task.is_summary
            and task.early_start
            and task.early_finish
        ):
            return task
    return None


def unscheduled_service_items(
    schedule: ProjectSchedule,
    items: list[BudgetItem],
) -> list[BudgetItem]:
    from pricing.budget.ppd_layout import ROW_TYPE_SERVICO

    return [
        item
        for item in items
        if item.row_type == ROW_TYPE_SERVICO
        and not item.metadata.get("is_memory_row")
        and not _leaf_task_for_service(schedule, item.row_id)
    ]


def _service_physical_weight(item: BudgetItem) -> float:
    qty = max(0.0, float(item.quantity or 0))
    return qty if qty > 0 else 1.0


def _service_financial_cost(item: BudgetItem) -> float:
    return float(item.effective_total())


def _leaf_tasks_for_curve(
    schedule: ProjectSchedule,
    visible_tasks: list[ScheduleTask] | None = None,
) -> list[ScheduleTask]:
    all_leaves = [
        t for t in schedule.tasks if not t.is_summary and t.early_start and t.early_finish
    ]
    if not visible_tasks:
        return all_leaves

    etapa_only = all(t.is_summary and t.row_type == "ETAPA" for t in visible_tasks)
    if not etapa_only:
        ids = {t.task_id for t in visible_tasks}
        return [t for t in all_leaves if t.task_id in ids]

    result: list[ScheduleTask] = []
    for etapa in visible_tasks:
        if not (etapa.early_start and etapa.early_finish):
            continue
        dur = max(1, _days_between(etapa.early_start, etapa.early_finish) + 1)
        result.append(
            ScheduleTask(
                task_id=etapa.task_id,
                budget_row_id=etapa.budget_row_id,
                budget_code=etapa.budget_code,
                name=etapa.name,
                row_type=etapa.row_type,
                parent_code=etapa.parent_code,
                duration_days=dur,
                is_summary=etapa.is_summary,
                manual_start=etapa.manual_start,
                early_start=etapa.early_start,
                early_finish=etapa.early_finish,
                late_start=etapa.late_start,
                late_finish=etapa.late_finish,
                total_float_days=etapa.total_float_days,
                is_critical=etapa.is_critical,
            )
        )
    return result


def _weight_and_cost(
    task: ScheduleTask,
    rows: list[BudgetItem],
) -> tuple[float, float]:
    from pricing.budget.ppd_layout import ROW_TYPE_SERVICO

    if task.is_summary:
        prefix = f"{task.budget_code}."
        children = [
            r
            for r in rows
            if r.row_type == ROW_TYPE_SERVICO
            and not r.metadata.get("is_memory_row")
            and (r.code == task.budget_code or str(r.code).startswith(prefix))
        ]
        weight = 0.0
        cost = 0.0
        for row in children:
            qty = max(0.0, float(row.quantity or 0))
            weight += qty if qty > 0 else 1.0
            cost += _service_financial_cost(row)
        if weight <= 0:
            weight = float(len(children) or 1)
        return weight, cost

    row = _row_for_task(rows, task)
    qty = max(0.0, float(row.quantity or 0)) if row else 0.0
    weight = qty if qty > 0 else 1.0
    cost = _service_financial_cost(row) if row else 0.0
    return weight, cost


def build_schedule_curves_by_month(
    schedule: ProjectSchedule,
    rows: list[BudgetItem],
    visible_tasks: list[ScheduleTask] | None = None,
) -> tuple[list[MonthBucket], float, float]:
    project_start = schedule.project_start
    project_end = schedule.project_end or project_start
    month_count = count_months(project_start, project_end)
    leaves = _leaf_tasks_for_curve(schedule, visible_tasks)

    total_physical_weight = 0.0
    total_financial = 0.0

    for task in leaves:
        weight, cost = _weight_and_cost(task, rows)
        total_physical_weight += weight
        total_financial += cost

    orphans = unscheduled_service_items(schedule, rows)
    orphan_weight = 0.0
    orphan_cost = 0.0
    for row in orphans:
        orphan_weight += _service_physical_weight(row)
        orphan_cost += _service_financial_cost(row)
    total_physical_weight += orphan_weight
    total_financial += orphan_cost

    if total_physical_weight <= 0:
        total_physical_weight = float(len(leaves) or len(orphans) or 1)

    months: list[MonthBucket] = []
    physical_cum = 0.0
    financial_cum = 0.0
    orphans_allocated = False

    for m in range(month_count):
        m_start = _month_start_date(project_start, m)
        m_end = _month_end_date(m_start)
        period_start = _format_iso(m_start)
        period_end = _format_iso(m_end)
        clip = _clip_to_project(period_start, period_end, project_start, project_end)
        if not clip:
            continue

        physical_month = 0.0
        financial_month = 0.0

        for task in leaves:
            weight, cost = _weight_and_cost(task, rows)
            dur = max(1, task.duration_days)
            overlap = _overlap_days(
                task.early_start or period_start,
                task.early_finish or period_end,
                clip[0],
                clip[1],
            )
            if overlap <= 0:
                continue
            physical_month += (weight * overlap) / dur
            financial_month += (cost * overlap) / dur

        if not orphans_allocated and (orphan_weight > 0 or orphan_cost > 0):
            physical_month += orphan_weight
            financial_month += orphan_cost
            orphans_allocated = True

        physical_monthly_pct = (physical_month / total_physical_weight) * 100
        physical_cum = min(100.0, physical_cum + physical_monthly_pct)
        financial_cum += financial_month

        months.append(
            MonthBucket(
                month_index=len(months),
                label=month_label(m_start),
                month_start_iso=period_start,
                month_end_iso=period_end,
                physical_monthly_pct=physical_monthly_pct,
                physical_cumulative_pct=physical_cum,
                financial_monthly=financial_month,
                financial_cumulative=financial_cum,
            )
        )

    return months, total_financial, total_physical_weight

File: frontend/lib/schedule-curves.ts
import type { BudgetRow, ProjectSchedule, ScheduleTask } from "@/types/api";

export interface MonthBucket {
  monthIndex: number;
  label: string;
  monthStartIso: string;
  monthEndIso: string;
  physicalMonthlyPct: number;
  physicalCumulativePct: number;
  financialMonthly: number;
  financialCumulative: number;
}

/** @deprecated use MonthBucket — kept for compat during migration */
export interface WeekBucket {
  weekIndex: number;
  label: string;
  weekStartIso: string;
  physicalWeeklyPct: number;
  physicalCumulativePct: number;
  financialWeekly: number;
  financialCumulative: number;
}

function parseDate(iso: string): Date {
  const [y, m, d] = iso.slice(0, 10).split("-").map(Number);
  return new Date(y, m - 1, d);
}

function daysBetween(start: string, end: string): number {
  const a = parseDate(start);
  const b = parseDate(end);
  return Math.round((b.getTime() - a.getTime()) / 86400000);
}

function addDays(iso: string, days: number): string {
  const d = parseDate(iso);
  d.setDate(d.getDate() + days);
  return d.toISOString().slice(0, 10);
}

function overlapDays(
  rangeStart: string,
  rangeEnd: string,
  periodStart: string,
  periodEnd: string
): number {
  const rs = parseDate(rangeStart).getTime();
  const re = parseDate(rangeEnd).getTime();
  const ps = parseDate(periodStart).getTime();
  const pe = parseDate(periodEnd).getTime();
  const start = Math.max(rs, ps);
  const end = Math.min(re, pe);
  if (end < start) return 0;
  return Math.round((end - start) / 86400000) + 1;
}

function monthStartDate(projectStart: string, monthIndex: number): Date {
  const s = parseDate(projectStart);
  return new Date(s.getFullYear(), s.getMonth() + monthIndex, 1);
}

function monthEndDate(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth() + 1, 0);
}

function formatIso(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

export function countMonths(projectStart: string, projectEnd: string): number {
  const s = parseDate(projectStart);
  const e = parseDate(projectEnd);
  return Math.max(1, (e.getFullYear() - s.getFullYear()) * 12 + (e.getMonth() - s.getMonth()) + 1);
}

export function monthLabel(d: Date): string {
  return d.toLocaleDateString("pt-BR", { month: "short", year: "2-digit" });
}

/** Exibe data ISO (YYYY-MM-DD) no padrão brasileiro dd/mm/aaaa */
export function formatIsoDateBr(iso?: string | null): string {
  if (!iso) return "—";
  const [y, m, d] = iso.slice(0, 10).split("-");
  if (!y || !m || !d) return iso;
  return `${d.padStart(2, "0")}/${m.padStart(2, "0")}/${y}`;
}

export interface TimelineWeekColumn {
  weekIndex: number;
  label: string;
  dateLabel: string;
  weekStartIso: string;
  weekEndIso: string;
  monthKey: string;
}

export interface TimelineMonthGroup {
  monthIndex: number;
  label: string;
  weekCount: number;
  monthKey: string;
}

export function countProjectWeeks(projectStart: string, projectEnd: string): number {
  const totalDays = Math.max(1, daysBetween(projectStart, projectEnd || projectStart) + 1);
  return Math.ceil(totalDays / 7);
}

export function buildProjectTimeline(
  projectStart: string,
  projectEnd: string
): {
  weeks: TimelineWeekColumn[];
  monthGroups: TimelineMonthGroup[];
  weekCount: number;
  totalDays: number;
} {
  const end = projectEnd || projectStart;
  const totalDays = Math.max(1, daysBetween(projectStart, end) + 1);
  const weekCount = Math.ceil(totalDays / 7);
  const weeks: TimelineWeekColumn[] = [];

  for (let w = 0; w < weekCount; w++) {
    const weekStart = addDays(projectStart, w * 7);
    const weekEnd = addDays(projectStart, Math.min(w * 7 + 6, totalDays - 1));
    const startDate = parseDate(weekStart);
    const monthKey = `${startDate.getFullYear()}-${String(startDate.getMonth() + 1).padStart(2, "0")}`;
    weeks.push({
      weekIndex: w,
      label: `S${w + 1}`,
      dateLabel: formatIsoDateBr(weekStart),
      weekStartIso: weekStart,
      weekEndIso: weekEnd,
      monthKey,
    });
  }

  const monthGroups: TimelineMonthGroup[] = [];
  for (const week of weeks) {
    const last = monthGroups[monthGroups.length - 1];
    if (!last || last.monthKey !== week.monthKey) {
      monthGroups.push({
        monthIndex: monthGroups.length,
        label: monthLabel(parseDate(week.weekStartIso)),
        weekCount: 1,
        monthKey: week.monthKey,
      });
    } else {
      last.weekCount += 1;
    }
  }

  return { weeks, monthGroups, weekCount, totalDays };
}

export function weeksOverlappingPeriod(
  weeks: TimelineWeekColumn[],
  periodStart: string,
  periodEnd: string
): number {
  let n = 0;
  for (const w of weeks) {
    if (overlapDays(w.weekStartIso, w.weekEndIso, periodStart, periodEnd) > 0) n += 1;
  }
  return n;
}

function clipToProject(
  periodStart: string,
  periodEnd: string,
  projectStart: string,
  projectEnd: string
): { start: string; end: string } | null {
  const end = projectEnd || projectStart;
  const start = periodStart < projectStart ? projectStart : periodStart;
  const finish = periodEnd > end ? end : periodEnd;
  if (start > finish) return null;
  return { start, end: finish };
}

/** Larguras das colunas mensais alinhadas ao timeline (soma = totalTimelinePx). */
export function monthTimelineWidths(
  months: MonthBucket[],
  projectStart: string,
  projectEnd: string,
  totalTimelinePx: number
): number[] {
  if (months.length === 0 || totalTimelinePx <= 0) return [];

  const end = projectEnd || projectStart;
  const dayWeights = months.map((m) => {
    const clip = clipToProject(m.monthStartIso, m.monthEndIso, projectStart, end);
    if (!clip) return 0;
    return daysBetween(clip.start, clip.end) + 1;
  });

  const weightSum = dayWeights.reduce((a, b) => a + b, 0) || 1;
  const widths = dayWeights.map((d) =>
    d <= 0 ? 0 : Math.max(1, Math.floor((d / weightSum) * totalTimelinePx))
  );

  let sum = widths.reduce((a, b) => a + b, 0);
  for (let i = widths.length - 1; sum < totalTimelinePx && i >= 0; i--) {
    if (dayWeights[i] > 0) {
      widths[i] += 1;
      sum += 1;
    }
  }
  for (let i = widths.length - 1; sum > totalTimelinePx && i >= 0; i--) {
    if (widths[i] > 1) {
      widths[i] -= 1;
      sum -= 1;
    }
  }

  return widths;
}

function rowForTask(rows: BudgetRow[], task: ScheduleTask): BudgetRow | undefined {
  return rows.find((r) => r.row_id === task.budget_row_id);
}

function leafTaskForService(schedule: ProjectSchedule, rowId: string): ScheduleTask | undefined {
  return schedule.tasks.find(
    (t) => t.budget_row_id === rowId && !t.is_summary && t.early_start && t.early_finish
  );
}

function servicePhysicalWeight(row: BudgetRow): number {
  const qty = Math.max(0, row.quantity ?? 1);
  return qty > 0 ? qty : 1;
}

function serviceFinancialCost(row: BudgetRow): number {
  return row.total_effective ?? row.total_price ?? 0;
}

/** Serviços (S) sem tarefa folha com datas no cronograma */
export function unscheduledServiceRows(
  schedule: ProjectSchedule,
  rows: BudgetRow[]
): BudgetRow[] {
  return rows.filter(
    (r) =>
      r.row_type === "S" && !r.is_memory_row && !leafTaskForService(schedule, r.row_id)
  );
}

function leafTasksForCurve(
  schedule: ProjectSchedule,
  visibleTasks?: ScheduleTask[]
): ScheduleTask[] {
  const allLeaves = schedule.tasks.filter(
    (t) => !t.is_summary && t.early_start && t.early_finish
  );
  if (!visibleTasks?.length) return allLeaves;

  const etapaOnly = visibleTasks.every((t) => t.is_summary && t.row_type === "ETAPA");
  if (!etapaOnly) {
    const ids = new Set(visibleTasks.map((t) => t.task_id));
    return allLeaves.filter((t) => ids.has(t.task_id));
  }

  return visibleTasks
    .filter((t) => t.early_start && t.early_finish)
    .map((etapa) => ({
      ...etapa,
      duration_days: Math.max(
        1,
        daysBetween(etapa.early_start!, etapa.early_finish!) + 1
      ),
    }));
}

function weightAndCost(
  task: ScheduleTask,
  rows: BudgetRow[],
  _etapaMode?: boolean
): { weight: number; cost: number } {
  if (task.is_summary) {
    const prefix = `${task.budget_code}.`;
    const children = rows.filter(
      (r) =>
        r.row_type === "S" &&
        !r.is_memory_row &&
        (r.code === task.budget_code || r.code.startsWith(prefix))
    );
    let weight = 0;
    let cost = 0;
    for (const row of children) {
      const qty = Math.max(0, row.quantity ?? 1);
      weight += qty > 0 ? qty : 1;
      cost += row.total_effective ?? row.total_price ?? 0;
    }
    if (weight <= 0) weight = children.length || 1;
    return { weight, cost };
  }

  const row = rowForTask(rows, task);
  const qty = Math.max(0, row?.quantity ?? 1);
  const weight = qty > 0 ? qty : 1;
  const cost = row?.total_effective ?? row?.total_price ?? 0;
  return { weight, cost };
}

export interface TaskScheduleMetrics {
  cost: number;
  financialPct: number;
  physicalPct: number;
}

export function computeTaskMetrics(
  schedule: ProjectSchedule,
  rows: BudgetRow[],
  visibleTasks: ScheduleTask[]
): {
  byTaskId: Map<string, TaskScheduleMetrics>;
  totalFinancial: number;
  totalPhysicalWeight: number;
} {
  const etapaOnly =
    visibleTasks.length > 0 &&
    visibleTasks.every((t) => t.is_summary && t.row_type === "ETAPA");

  const leaves = leafTasksForCurve(schedule, visibleTasks);
  let totalPhysicalWeight = 0;
  let totalFinancial = 0;

  for (const task of leaves) {
    const { weight, cost } = weightAndCost(task, rows, etapaOnly);
    totalPhysicalWeight += weight;
    totalFinancial += cost;
  }
  if (totalPhysicalWeight <= 0) totalPhysicalWeight = leaves.length || 1;

  const byTaskId = new Map<string, TaskScheduleMetrics>();
  for (const task of visibleTasks) {
    const { weight, cost } = weightAndCost(task, rows, etapaOnly);
    byTaskId.set(task.task_id, {
      cost,
      financialPct: totalFinancial > 0 ? (cost / totalFinancial) * 100 : 0,
      physicalPct: totalPhysicalWeight > 0 ? (weight / totalPhysicalWeight) * 100 : 0,
    });
  }

  return { byTaskId, totalFinancial, totalPhysicalWeight };
}

/** Soma % físico apenas das folhas da visão (evita dupla contagem etapa + serviços). */
export function totalPhysicalPctForView(
  schedule: ProjectSchedule,
  visibleTasks: ScheduleTask[],
  byTaskId: Map<string, TaskScheduleMetrics>
): number {
  const leaves = leafTasksForCurve(schedule, visibleTasks);
  const total = leaves.reduce(
    (sum, task) => sum + (byTaskId.get(task.task_id)?.physicalPct ?? 0),
    0
  );
  return Math.min(100, total);
}

export function buildScheduleCurvesByMonth(
  schedule: ProjectSchedule,
  rows: BudgetRow[],
  visibleTasks?: ScheduleTask[]
): { months: MonthBucket[]; totalFinancial: number; totalPhysicalWeight: number } {
  const projectStart = schedule.project_start;
  const projectEnd = schedule.project_end || projectStart;
  const monthCount = countMonths(projectStart, projectEnd);
  const leaves = leafTasksForCurve(schedule, visibleTasks);
  const etapaMode = Boolean(
    visibleTasks?.length &&
      visibleTasks.every((t) => t.is_summary && t.row_type === "ETAPA")
  );

  let totalPhysicalWeight = 0;
  let totalFinancial = 0;

  for (const task of leaves) {
    const { weight, cost } = weightAndCost(task, rows, etapaMode);
    totalPhysicalWeight += weight;
    totalFinancial += cost;
  }

  const orphans = unscheduledServiceRows(schedule, rows);
  let orphanWeight = 0;
  let orphanCost = 0;
  for (const row of orphans) {
    orphanWeight += servicePhysicalWeight(row);
    orphanCost += serviceFinancialCost(row);
  }
  totalPhysicalWeight += orphanWeight;
  totalFinancial += orphanCost;

  if (totalPhysicalWeight <= 0) totalPhysicalWeight = leaves.length || orphans.length || 1;

  const months: MonthBucket[] = [];
  let physicalCum = 0;
  let financialCum = 0;
  let orphansAllocated = false;

  for (let m = 0; m < monthCount; m++) {
    const mStart = monthStartDate(projectStart, m);
    const mEnd = monthEndDate(mStart);
    const periodStart = formatIso(mStart);
    const periodEnd = formatIso(mEnd);
    const clip = clipToProject(periodStart, periodEnd, projectStart, projectEnd);
    if (!clip) continue;

    let physicalMonth = 0;
    let financialMonth = 0;

    for (const task of leaves) {
      const { weight, cost } = weightAndCost(task, rows, etapaMode);
      const dur = Math.max(1, task.duration_days);
      const overlap = overlapDays(
        task.early_start!,
        task.early_finish!,
        clip.start,
        clip.end
      );
      if (overlap <= 0) continue;
      physicalMonth += (weight * overlap) / dur;
      financialMonth += (cost * overlap) / dur;
    }

    if (!orphansAllocated && (orphanWeight > 0 || orphanCost > 0)) {
      physicalMonth += orphanWeight;
      financialMonth += orphanCost;
      orphansAllocated = true;
    }

    const physicalMonthlyPct = (physicalMonth / totalPhysicalWeight) * 100;
    physicalCum = Math.min(100, physicalCum + physicalMonthlyPct);
    financialCum += financialMonth;

    months.push({
      monthIndex: months.length,
      label: monthLabel(mStart),
      monthStartIso: periodStart,
      monthEndIso: periodEnd,
      physicalMonthlyPct,
      physicalCumulativePct: physicalCum,
      financialMonthly: financialMonth,
      financialCumulative: financialCum,
    });
  }

  return { months, totalFinancial, totalPhysicalWeight };
}

/** @deprecated use buildScheduleCurvesByMonth */
export function buildScheduleCurves(
  schedule: ProjectSchedule,
  rows: BudgetRow[]
): { weeks: WeekBucket[]; totalFinancial: number; totalPhysicalWeight: number } {
  const { months, totalFinancial, totalPhysicalWeight } = buildScheduleCurvesByMonth(
    schedule,
    rows
  );
  const weeks: WeekBucket[] = months.map((m) => ({
    weekIndex: m.monthIndex,
    label: m.label,
    weekStartIso: m.monthStartIso,
    physicalWeeklyPct: m.physicalMonthlyPct,
    physicalCumulativePct: m.physicalCumulativePct,
    financialWeekly: m.financialMonthly,
    financialCumulative: m.financialCumulative,
  }));
  return { weeks, totalFinancial, totalPhysicalWeight };
}

/** @deprecated use countProjectWeeks */
export function countWeeks(projectStart: string, projectEnd: string): number {
  return countProjectWeeks(projectStart, projectEnd);
}

export function fmtCurrency(n: number): string {
  if (n >= 1_000_000) return `R$ ${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `R$ ${(n / 1_000).toFixed(0)}k`;
  return n.toLocaleString("pt-BR", { style: "currency", currency: "BRL", maximumFractionDigits: 0 });
}

export function fmtCurrencyCompact(n: number): string {
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(0)}k`;
  return n.toLocaleString("pt-BR", { maximumFractionDigits: 0 });
}

export function fmtPct(n: number): string {
  if (n >= 10) return `${n.toFixed(1)}%`;
  if (n >= 1) return `${n.toFixed(1)}%`;
  return n > 0 ? `${n.toFixed(2)}%` : "0%";
}

export type ScheduleViewMode = "etapas" | "completo";

export function filterScheduleTasks(
  tasks: ScheduleTask[],
  mode: ScheduleViewMode
): ScheduleTask[] {
  if (mode === "completo") return tasks;
  return tasks.filter((t) => t.is_summary && t.row_type === "ETAPA");
}

export { daysBetween, addDays, parseDate };
